auth/auth/auth.py

```diff
@@ -0,0 +1,233 @@
+
+
+@routes.get('/healthcheck')
+async def get_healthcheck(request):  # pylint: disable=W0613
+    return web.Response()
+
+
+@routes.get('')
+@routes.get('/')
+async def get_index(request):  # pylint: disable=unused-argument
+    return aiohttp.web.HTTPFound('/login')
+
+
+@routes.get('/login')
+@web_maybe_authenticated_user
+async def login(request, userdata):
+    next = request.query.get('next', deploy_config.external_url('notebook2', ''))
+    if userdata:
+        return aiohttp.web.HTTPFound(next)
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'))
+
+    authorization_url, state = flow.authorization_url(
+        access_type='offline',
+        include_granted_scopes='true')
+
+    session = await aiohttp_session.get_session(request)
+    session['state'] = state
+    session['next'] = next
+
+    return aiohttp.web.HTTPFound(authorization_url)
+
+
+@routes.get('/oauth2callback')
+async def callback(request):
+    session = await aiohttp_session.get_session(request)
+    if 'state' not in session:
+        raise web.HTTPUnauthorized()
+    state = session['state']
+
+    flow = get_flow(deploy_config.external_url('auth', '/oauth2callback'), state=state)
+
+    try:
+        flow.fetch_token(code=request.query['code'])
+        token = google.oauth2.id_token.verify_oauth2_token(
+            flow.credentials.id_token, google.auth.transport.requests.Request())
+        id = token['sub']
+    except Exception:
+        log.exception('oauth2 callback: could not fetch and verify token')
+        raise web.HTTPUnauthorized()
+
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('SELECT * from user_data where user_id = %s;', f'google-oauth2|{id}')
+            users = await cursor.fetchall()
+
+    if len(users) != 1:
+        raise web.HTTPUnauthorized()
+    user = users[0]
+
+    session_id = await create_session(dbpool, user['id'])
+
+    del session['state']
+    session['session_id'] = session_id
+    next = session.pop('next')
+    return aiohttp.web.HTTPFound(next)
+
+
+@routes.post('/logout')
+@web_maybe_authenticated_user
+async def logout(request, userdata):
+    if not userdata:
+        return web.HTTPFound(deploy_config.external_url('notebook2', ''))
+
+    dbpool = request.app['dbpool']
+    session_id = userdata['session_id']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('DELETE FROM sessions WHERE session_id = %s;', session_id)
+
+    session = await aiohttp_session.get_session(request)
+    if 'session_id' in session:
+        del session['session_id']
+
+    return web.HTTPFound(deploy_config.external_url('notebook2', ''))
+
+
+@routes.get('/api/v1alpha/login')
+async def rest_login(request):
+    callback_port = request.query['callback_port']
+
+    flow = get_flow(f'http://127.0.0.1:{callback_port}/oauth2callback')
```

Comment:
In this scenario it looks like the user is running auth.py locally? If so, that may not be something we allow to happen at all.

Also, is the callback always going to be localhost?

Reply:
And should this be `http://127.0.0.1:{callback_port}/api/v1alpha/oauth2callback`?

Reply:
> running auth.py locally 

Which auth.py do you mean here?  It's not running auth/auth.py service, the login command runs its own server to accept the callback.  FYI, I double-checked and gcloud uses an http (non-secure) callback:

```
$ gcloud auth login
Your browser has been opened to visit:

    https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8085%2F&...
```

No, it's calling the server in login.py:

```
@routes.get('/oauth2callback')
async def callback(request):
    ...
```

Since those two pieces of code are coupled, I don't think versioning is necessary.

Reply:
I've read up on this a bit more, and it seems that the way google protects against this depends on the presence of a client secret.

The solution for clients that don't have the secret is something called PKCE, in which a dynamic secret is created by the client (here the requestor), with each request for an authorization code.

The secret is a long random number, which is then hashed by SHA256. This is sent with the authorization url request. The authorization server logs this hashed secret and returns a valid url. The client then uses this url?code_verifier=sha256(secret). The token-generating server, looks up the hashed secret stored for this authorization url, and verifies that the codes match.

This way, if an attacker is able to set itself as the callback for authorization_url, it cannot use that authorization_url to grab a valid token.

This attack requires a long list of preconditions, which may seem to us ridiculous, but I feel that if a standard method exists to protect against this, we should use it.

https://tools.ietf.org/html/rfc7636
 +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    | End Device (e.g., Smartphone)  |
    |                                |
    | +-------------+   +----------+ | (6) Access Token  +----------+
    | |Legitimate   |   | Malicious|<--------------------|          |
    | |OAuth 2.0 App|   | App      |-------------------->|          |
    | +-------------+   +----------+ | (5) Authorization |          |
    |        |    ^          ^       |        Grant      |          |
    |        |     \         |       |                   |          |
    |        |      \   (4)  |       |                   |          |
    |    (1) |       \  Authz|       |                   |          |
    |   Authz|        \ Code |       |                   |  Authz   |
    | Request|         \     |       |                   |  Server  |
    |        |          \    |       |                   |          |
    |        |           \   |       |                   |          |
    |        v            \  |       |                   |          |
    | +----------------------------+ |                   |          |
    | |                            | | (3) Authz Code    |          |
    | |     Operating System/      |<--------------------|          |
    | |         Browser            |-------------------->|          |
    | |                            | | (2) Authz Request |          |
    | +----------------------------+ |                   +----------+
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+

This may be a little weird to use because we construct a single instance of Flow (we would need one per api/ v1alpha/login request).

I thought we could use an inverted process, or if we solve the above, as an additional layer of protection: we use a dynamic RSA encryption. Every time the cli client calls hailtcl auth login, the login.py server creates a public/private key pair. It sends the public key along with the request for the authorization_url (to api server), the api server encrypts the response using this public key, and only the possessor of the private key can decrypt that response. Seems safer.

Reply:
In this vein, apparently item recommends PKCE for all clients, native or not: 

https://tools.ietf.org/html/draft-ietf-oauth-security-topics-11#section-2.1.1
```
  Note: although PKCE so far was recommended as a mechanism to protect
   native apps, this advice applies to all kinds of OAuth clients,
   including web applications.
```

Reply:
I'm having trouble correlating your references with the Google flow code.  Here is my best guess: I think we have a client secret (which Google calls state).  Google's `Flow.authorization_url` returns the authorization url along with a state.  I think state is a random token.  Here is what the Google docs say about the state:

    The state is used when completing the flow to verify that the request originated from your application.

https://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.Flow.authorization_url

The state is returned by auth/api/v1alpha/login to the client (encrypted).  Then the user does the login flow with Google using the authorization url but NOT the state, and the client gets an (local, unencrypted) callback from the browser with the "code".  The client then makes an (encrypted) callback to auth/api/v1alpha/oauth2callback with the state (which was never exposed) and the code.  /oauth2callback then completes the verification with Google using both the state and the code by calling `Flow.get_token` to get the final JWT.  Attacker who captures the code cannot complete the oauth2 flow without the state.  I have verified this (by not propagating the state in /oauth2callback.)

 > This may be a little weird to use because we construct a single instance of Flow (we would need one per api/ v1alpha/login request).

No, we create a Flow per call to auth/api/v1alpha/login by calling `get_flow`.

Reply:
> No, we create a Flow per call to auth/api/v1alpha/login by calling get_flow.

Oops, of course. So we could use this PKCE approach.

code_verifier = kwargs.pop('code_verifier', None)
        autogenerate_code_verifier = kwargs.pop(
            'autogenerate_code_verifier', None)

> The client then makes an (encrypted) callback to auth/api/v1alpha/oauth2callback with the state (which was never exposed) and the code. /oauth2callback then completes the verification with Google using both the state and the code by calling Flow.get_token to get the final JWT

Wait, on the cli side, no call to /auth/api/v1alpha/oauth2callback is made right? login.py gets a response with the authorization url, which is set to /oauth2callback on the cli's local server.

> The state is returned by auth/api/v1alpha/login to the client (encrypted).

From my understanding the PKCE implementation is orthogonal. 

The state secret generated by [requests_outhlib (used by google flow)](https://requests-oauthlib.readthedocs.io/en/stable/api.html#requests_oauthlib.OAuth2Session.authorization_url) is used for CSRF protection (i.e an attacker who could hijack a session, but not read the contents of the cookie that contains the state because it's httpOnly couldn't use the authorization_url).

PKCE is like inverse CSRF. The client creates the secret, and never send it out (except encrypted and over https), so even if the authorization_url is intercepted, it is useless.

In the google flow code, we distinguish between state and `code_verifier`:

```python
def authorization_url(self, **kwargs):
        """Generates an authorization URL.

        This is the first step in the OAuth 2.0 Authorization Flow. The user's
        browser should be redirected to the returned URL.

        This method calls
        :meth:`requests_oauthlib.OAuth2Session.authorization_url`
        and specifies the client configuration's authorization URI (usually
        Google's authorization server) and specifies that "offline" access is
        desired. This is required in order to obtain a refresh token.

        Args:
            kwargs: Additional arguments passed through to
                :meth:`requests_oauthlib.OAuth2Session.authorization_url`

        Returns:
            Tuple[str, str]: The generated authorization URL and state. The
                user must visit the URL to complete the flow. The state is used
                when completing the flow to verify that the request originated
                from your application. If your application is using a different
                :class:`Flow` instance to obtain the token, you will need to
                specify the ``state`` when constructing the :class:`Flow`.
        """
        kwargs.setdefault('access_type', 'offline')
        if self.autogenerate_code_verifier:
            chars = ascii_letters+digits+'-._~'
            rnd = SystemRandom()
            random_verifier = [rnd.choice(chars) for _ in range(0, 128)]
            self.code_verifier = ''.join(random_verifier)

        if self.code_verifier:
            code_hash = hashlib.sha256()
            code_hash.update(str.encode(self.code_verifier))
            unencoded_challenge = code_hash.digest()
            b64_challenge = urlsafe_b64encode(unencoded_challenge)
            code_challenge = b64_challenge.decode().split('=')[0]
            kwargs.setdefault('code_challenge', code_challenge)
            kwargs.setdefault('code_challenge_method', 'S256')
        url, state = self.oauth2session.authorization_url(
            self.client_config['auth_uri'], **kwargs)

        return url, state
```

Reply:
Tracked use; so from here, google calls https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_session.py#L154, which passes through the code_challenge and code_challenge_method to https://github.com/oauthlib/oauthlib, which uses it in their PKCE implementation (and separately from the CSRF stuff).